README.md
# `nemid`

[![Conventional Commits](https://img.shields.io/badge/Conventional%20Commits-1.0.0-yellow.svg)](https://conventionalcommits.org)

> Node.js module for NemID authentication and signing

## Install

```sh
npm install @tavexit/nemid
```

## Usage

Server:

```js
const { NemID } = require('@tavexit/nemid')
const fs = require('fs')
const path = require('path')
const crypto = require('crypto')

// Example origin
const ORIGIN = 'https://localhost:8080'

// Paths to keys and certificates
const cert = path.join(__dirname, 'cert.der')
const key = path.join(__dirname, 'key.pem')
const issuer = path.join(__dirname, 'issuer.der')

const nemid = new NemID({
  clientKey: crypto.createPrivateKey({
    key: fs.readFileSync(key),
    format: 'pem',
    type: 'pkcs1'
  }),
  clientCert: fs.readFileSync(cert),
  serverCA: fs.readFileSync(issuer)
})

server.get('/authenticate', (req, res) => {
  return res.send(nemid.authenticate({ origin: ORIGIN }))
})

server.post('/authenticate/verify', (req, res) => {
  const response = req.body

  nemid.verifyAuthenticate(response)
    .then((userInfo) => {
      if (userInfo === false) {
        return res.send({ success: false })
      }

      // Can do stuff with userInfo now
      console.log(userInfo.serialNumber) // contains the user PID

      return res.send({ success: true })
    })
    .catch((err) => {
      // err is a NemIDError
      // Log the cause of the error on the server somehow
      console.error(err)
      // And send a user message to the client
      return res.send(err.userMessage.da)
    })
  })
})
```

Client:

```js
const { getNemIDAuthContext } = require('@tavexit/nemid')
const { data: parameters } = await get('http://localhost:8080/authenticate')

const context = getNemIDAuthContext(parameters)

document.body.appendChild(context.element)

const result = await context.done;

const { data: success } = await post('http://localhost:8080/authenticate/verify', result)
```

## Server API

### `const nemid = new NemID({ spid, clientKey, clientCert, serverCA, env = NemID.TEST })`

Create a new NemID instance. Takes 5 arguments:

* `spid` is the Service Provider ID, provided by Nets at registration.
* `clientKey` must be the client private key provided by Nets. See the section
  below for how to extract the key from the service provider bundle.
* `clientCert` must be the client certificate, including intermediate certs.
  See the section below for how to extract this from the service provider bundle.
* `serverCA` must be the Nets root certificate. This can be downloaded from
  NemIDs website.
* `env` is an object of endpoints for the different environments. Comes with
  `NemID.TEST` and `NemID.PROD` built in.

### `const parameters = nemid.authenticate({ origin })`

Create a new authentication attempt, by generating the appropriate parameters to
be passed to the client side script.

### `const userInfo = await nemid.verifyAuthenticate(response)`

Verify the response from the NemID applet. May throw an error if the
result was malformed or an error code from NemID. Otherwise returns `false`
for an invalid attempt or a object describing the authenticated user.

### `const valid = nemid.matchCPR(pid, cpr)`

Verify whether `pid` and `cpr` refer to the same person.

## Client API

```ts
import { getNemIDAuthContext } from '@noeignite/nemid`;
```

### `const context = getNemIDAuthContext(parameters, prod = false)`

Initialise a new NemID authentication context with `parameters` from the Server
API and which origin frame to load. Defaults to test (`appletk.danid.dk`).
Returns a context containing `{ element, done }`.

### `const iframeElm = context.element`

`HTMLIFrameElement` that you can place in the DOM.

### `const result = await context.done`

Wait for the authentication to be done. Rejects in case of unexpected failure or
resolves when there's a reply from NemID (which might not be a successful auth).

### Styling

The `element` can be targeted by the CSS class `nemid-iframe`. Below is the
default styling from NemID:

```css
.nemid-iframe {
  border: 0;
  width: 320px;
  height: 480px;
}
```

## Working with `.p12`

To extract all certificates, and print them to `stdout`:

```sh
openssl pkcs12 -in bundle.p12 -nokeys
```

Find you company certificate, and copy `-----BEGIN CERTIFICATE-----` to
`-----END CERTIFICATE-----` inclusive and paste into a new file, eg `cert.pem`.
This file is now in PEM format, but Nets requires it to be in DER format. It can
be converted with:

```sh
openssl x509 -inform pem -outform der -text -in cert.pem -out cert.der
```

The remaining certificates from the first command were part of the certificate
chain and need to be placed in their own file. Start by copying the sections
including the start and end markers to their own file, eg. `issuer.pem`.
Then convert them to DER format with:

```sh
openssl x509 -inform pem -outform der -text -in issuer.pem -out issuer.der
```

To extract the private key:

```sh
openssl pkcs12 -in bundle.p12 -nocerts -nodes
```

And copy the output including the start and end markers
`-----BEGIN PRIVATE KEY-----` and `-----END PRIVATE KEY-----` to `key.pem`.
This file does not need to be converted to DER format.

**Note** `-nodes` removes the passphrase from the private key. If you want to
keep this, remove the option and supply your passphrase to
`crypto.createPrivateKey`

## License

[ISC](LICENSE)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.11](https://github.com/tavexit/nemid/compare/v1.0.10...v1.0.11) (2022-04-14)

### [1.0.10](https://github.com/tavexit/nemid/compare/v1.0.9...v1.0.10) (2022-04-13)

### [1.0.9](https://github.com/tavexit/nemid/compare/v1.0.8...v1.0.9) (2022-04-13)

### [1.0.8](https://github.com/tavexit/nemid/compare/v1.0.7...v1.0.8) (2022-04-13)

### [1.0.8](https://github.com/NoA-Ignite-dk/nemid/compare/v1.0.7...v1.0.8) (2022-04-13)

### [1.0.5](https://github.com/NoA-Ignite-dk/nemid/compare/v1.0.4...v1.0.5) (2021-08-24)


### Bug Fixes

* access element child ([f08d946](https://github.com/NoA-Ignite-dk/nemid/commit/f08d946d96c0b604ce3aa524eee7e26eb486a2f9))

### [1.0.4](https://github.com/NoA-Ignite-dk/nemid/compare/v1.0.3...v1.0.4) (2021-08-24)


### Bug Fixes

* import xml parser ([ea7ca78](https://github.com/NoA-Ignite-dk/nemid/commit/ea7ca783a6901c702ac5b1de83aca718f949e1f2))

### [1.0.3](https://github.com/NoA-Ignite-dk/nemid/compare/v1.0.2...v1.0.3) (2021-08-20)

### [1.0.2](https://github.com/NoA-Ignite-dk/nemid/compare/v1.0.1...v1.0.2) (2021-08-20)

### [1.0.1](https://github.com/NoA-Ignite-dk/nemid/compare/v1.0.0...v1.0.1) (2021-06-22)

## 1.0.0 (2021-06-22)


src/index.ts
import compare from 'compare';
import assert from 'nanoassert';
import crypto from 'crypto';
import WebCrypto from 'node-webcrypto-ossl';
import * as XmlDSigJs from 'xmldsigjs';
import { PIDCPRRequest } from './pid-cpr-request';
import { Code, errors, NemIDError, NemIDErrorType } from './error';
export type { getNemIDAuthContext } from './browser';

// Unfortunately a singleton
XmlDSigJs.Application.setEngine('OpenSSL', new WebCrypto() as unknown as Crypto);

interface Parameters {
	clientflow: string;
	clientmode: string;
	timestamp: string;
	ORIGIN: string;
}

export interface SignedParameters {
	clientflow: string;
	clientmode: string;
	timestamp: string;
	ORIGIN: string;
	SP_CERT: string;
	PARAMS_DIGEST: string;
	DIGEST_SIGNATURE: string;
}

export interface UserInfo {
	C: string;
	O: string;
	CN: string;
	serialNumber: string;
}

interface NemIDParam {
	/** The Service Provider ID, provided by Nets at registration */
	spid: string,

	/** The client private key provided by Nets */
	clientKey: crypto.KeyObject,

	/** The client certificate, including intermediate certs. */
	clientCert: Buffer,

	/** Nets root certificate. This can be downloaded from NemIDs website. */
	serverCA: Buffer,

	/**
	 * An object of endpoints for the different environments.
	 * Comes with `NemID.TEST` and `NemID.PROD` built in.
	 * @default {pid:'pidws.pp.certifikat.dk'}
	 */
	env: { pid: string }
}

export class NemID {
	static PROD = { pid: PIDCPRRequest.PROD }
	static TEST = { pid: PIDCPRRequest.TEST }

	private _clientKey: crypto.KeyObject;
	private _clientCert: Buffer;
	private _serverCA: Buffer;
	private _lookup: PIDCPRRequest;

	constructor ({ spid, clientKey, clientCert, serverCA, env = NemID.TEST }: NemIDParam) {
		this._clientKey = clientKey;
		this._clientCert = clientCert;
		this._serverCA = serverCA;

		this._lookup = new PIDCPRRequest(spid, clientKey, clientCert.toString(), env.pid);
	}

	authenticate ({ origin }: { origin: string}) {
		const parameters: Parameters = {
			clientflow: 'OcesLogin2',
			clientmode: 'standard',
			timestamp: Buffer.from(new Date().toISOString().slice(0, -5).replace('T', ' ') + '+0000').toString('base64'),
			ORIGIN: origin
		};

		return NemID.signParameters({
			parameters,
			privateKey: this._clientKey,
			spCert: this._clientCert
		});
	}

	async verifyAuthenticate (nemIdResponse: string) {
		assert(typeof nemIdResponse === 'string', 'nemIdResponse must be string');

		const responseData = Buffer.from(nemIdResponse, 'base64').toString();
		const error = NemID.errorsByCode.get(responseData as Code);
		if (error != null) throw new NemIDError(error);

		// An RSA signature is well beyond 32 bytes
		if (responseData.length < 32) {
			const err = new NemIDError(NemID.errorsByCode.get('NODE001')!);
			err.cause += '. Input: ' + responseData;

			throw err;
		}

		let doc: Document;
		let signedXml: XmlDSigJs.SignedXml;
		try {
			doc = XmlDSigJs.Parse(responseData);
			const signature = doc.getElementsByTagNameNS('http://www.w3.org/2000/09/xmldsig#', 'Signature');

			signedXml = new XmlDSigJs.SignedXml(doc);
			signedXml.LoadXml(signature[0]);
		} catch (ex) {
			const err = new NemIDError(NemID.errorsByCode.get('NODE001')!);
			err.cause += '. Exception: ' + ex;

			throw err;
		}

		const isValid  = await signedXml.Verify();

		if (isValid === false) {
			return false;
		}

		const x509 = doc.getElementsByTagNameNS('http://www.w3.org/2000/09/xmldsig#', 'X509Data');

		const certs = Array.from(x509).flatMap(c => XmlDSigJs.KeyInfoX509Data.LoadXml(c).Certificates);

		const subjects = certs
			.map(c => c.Subject
				.split(', ')
				.reduce((o, kv) => {
					let [k, v] = kv.split('=');
					// remap this OID to the identifier name
					if (k === '2.5.4.5') k = 'serialNumber';

					o[k] = v;
					return o;
				}, {} as {[key: string]: string})
			);

		const user = subjects
			.find(c => c.serialNumber != null) as UserInfo | undefined;

		if (user == null) {
			return false;
		}

		return user;
	}

	async matchCPR (pid: string, cpr: string): Promise<boolean> {
		return this._lookup.match(pid, cpr);
	}

	static signParameters ({ parameters, privateKey, spCert }: { parameters: Parameters, privateKey: crypto.KeyObject, spCert: Buffer }): SignedParameters {
		const _keys = Object.keys(parameters).map(k => k.toLowerCase());
		assert(_keys.includes('sp_cert') === false);
		assert(_keys.includes('params_digest') === false);
		assert(_keys.includes('digest_signature') === false);

		const SP_CERT = spCert.toString('base64');
		const input = this.normalizedParameters({ ...parameters, SP_CERT});

		const signedParameters = {
			...parameters,
			SP_CERT: SP_CERT,
			PARAMS_DIGEST: crypto.createHash('sha256')
				.update(input)
				.digest('base64'),
			DIGEST_SIGNATURE: crypto.createSign('sha256')
				.update(input)
				.sign(privateKey, 'base64')
		};

		return signedParameters;
	}

	static normalizedParameters (parameters: Parameters & { SP_CERT: string }) {
		const keys = (
			Object
				.keys(parameters)
				.sort((a, b)  => compare(a.toLowerCase(), b.toLowerCase()))
		) as (keyof Parameters)[];

		return keys.reduce((sum: string, key: keyof Parameters) => {
			sum += key + parameters[key];
			return sum;
		}, '');
	}

	static NemIDError = NemIDError;
	static errorsByCode = errors.reduce((map, e) => {
		map.set(e.code, e);

		return map;
	}, new Map<Code, NemIDErrorType>())
}


example/server.js
const express = require('express');
const app = express();
const crypto = require('crypto');
const path = require('path');
const fs = require('fs');
const cert = path.join(__dirname, '../cert.der');
const key = path.join(__dirname, '../test.key');
const issuer = path.join(__dirname, '../issuer.der');
const { NemID } = require('@tavexit/nemid');

const nemid = new NemID({
	clientKey: crypto.createPrivateKey({
		key: fs.readFileSync(key),
		format: 'pem',
		type: 'pkcs1'
	}),
	clientCert: fs.readFileSync(cert),
	serverCA: fs.readFileSync(issuer)
});

app.use(require('cors')({
	origin: 'http://localhost:9966'
}));

app.use(express.json());

app.get('/authenticate', function (req, res) {
	const parameters = nemid.authenticate({ origin: 'http://localhost:9966' });

	res.send(parameters);
});

app.post('/authenticate/verify', function (req, res) {
	const response = req.body.content;

	nemid.verifyAuthenticate(response, function (err, userInfo) {
		if (err) {
			// err is a NemIDError
			// Log the cause of the error on the server somehow
			console.error(err);
			// And send a user message to the client
			return res.send(err.userMessage.da);
		}

		if (userInfo === false) {
			return res.send({ success: false });
		}

		// Can do stuff with userInfo now
		console.log(userInfo.serialNumber); // contains the user PID

		return res.send({ success: true });
	});
});

app.listen(8000);


example/example.js
const express = require('express');
const app = express();
const crypto = require('crypto');
const path = require('path');
const fs = require('fs');
const cert = path.join(__dirname, 'cert.der');
const key = path.join(__dirname, 'test.key');
const issuer = path.join(__dirname, 'issuer.der');
const { NemID } = require('@tavexit/nemid');

const nemid = new NemID({
	clientKey: crypto.createPrivateKey({
		key: fs.readFileSync(key),
		format: 'pem',
		type: 'pkcs1'
	}),
	clientCert: fs.readFileSync(cert),
	serverCA: fs.readFileSync(issuer)
});

app.use(express.json());

app.get('/', function (req, res) {
	res.setHeader('X-Frame-Options', 'deny');
	// res.setHeader('Content-Security-Policy', "")
	res.setHeader('X-Content-Type-Options', 'nosniff');
	res.setHeader('X-Download-Options', 'noopen');
	res.setHeader('Strict-Transport-Security', 'max-age=5184000; includeSubDomains; preload');

	const parameters = nemid.authenticate({ origin: 'http://localhost:8000' });

	res.send(`<!DOCTYPE html>
  <html lang="en" dir="ltr">
    <head>
      <meta charset="utf-8">
      <meta http-equiv="X-UA-Compatible" content="IE=edge">
      <meta name="viewport" content="width=device-width, initialscale=1.0, user-scalable=no">
      <title></title>
      <style>
        .nemid {
          border: 0;
          width: 320px;
          height: 480px;
        }
      </style>
      <script type="text/javascript">
        const NEMID_ORIGIN = 'https://appletk.danid.dk'

        function initNemid (parameters) {
          const elm = document.createElement('iframe')
          elm.id = 'nemid-' + Math.random().toString(32)
          elm.classList.add('nemid')
          elm.title = 'NemID'
          elm.allowfullscreen = true
          elm.src = \`\${NEMID_ORIGIN}/launcher/std/\${elm.id}\`

          window.addEventListener('message', handler)

          return { element: elm }

          function handler (ev) {
            if (ev.origin !== NEMID_ORIGIN) return
            if (ev.source !== elm.contentWindow) return

            // capture the event here
            ev.stopPropagation()

            var data = {}

            try {
              data = JSON.parse(ev.data)
            } catch (ex) {
              console.error(ev.data)
            }

            const { command, content } = data

            if (command === 'SendParameters') {
              const res = { command: 'parameters', content: JSON.stringify(parameters) }
              elm.contentWindow.postMessage(JSON.stringify(res), NEMID_ORIGIN)
              return false
            }

            if (command === 'changeResponseAndSubmit') {
              console.log('http://localhost:8000/', { content })
              // Terminal condition
              window.removeEventListener('message', handler)
              return false
            }
          }
        }
      </script>
    </head>
    <body>

      <script>
        const ctx = initNemid(${JSON.stringify(parameters)})
        document.body.appendChild(ctx.element)

        const ctx2 = initNemid(${JSON.stringify(parameters)})
        document.body.appendChild(ctx2.element)
      </script>
    </body>
  </html>`);
});

// eslint-disable-next-line @typescript-eslint/no-unused-vars
app.post('/', function (req, res) {
	nemid.verifyAuthenticate(req.body.content, console.log);
});

app.listen(8000);
